### PR TITLE
Use invokedynamic for structural calls, symbol literals, lambda ser.

### PR DESCRIPTION
**Pull request by retronym**
The previous encodings created static fields in the enclosing class
to host caches. However, this isn't an option once emit code in default
methods of interfaces, as Java interfaces don't allow static fields.

Luckily, we can emulate a static field by using invokedynamic: when
the call site is linked, our bootstrap methid can perform one-time
computation, and we can capture the result in the CallSite.

Review by @lrytz. 

I can't remember why I needed to change `visitHandle`, but I know that the ugliness there is a symptom of a somewhat ad-hoc representation of indy calls in our trees. Ideally, we'd be able to come up with something general purpose, so that a macro author could emit an arbitrary indy call, without needing special cases in the backend for each bootstrap method.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala [hunk 1]
```diff
@@ -1354,4 +1368,14 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
           ArrayBType(ObjectRef)),
         coreBTypes.jliCallSiteRef
       ).descriptor)
+
+  lazy val symbolLiteralBoostrapHandle =
+    new asm.Handle(asm.Opcodes.H_INVOKESTATIC,
+      "scala/runtime/SymbolLiteral", "bootstrap",
+      "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/String;)Ljava/lang/invoke/CallSite;")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala]: References to class names should go through a BType, see https://github.com/scala/scala/blob/fb22e2b0a73605d654c153e02d454e5cec21f355/src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala#L74-L78

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala [hunk 2]
```diff
@@ -1354,4 +1368,14 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
...
+
+  lazy val structuralCallSite_BoostrapHandle =
+    new asm.Handle(asm.Opcodes.H_INVOKESTATIC,
+      "scala/runtime/StructuralCallSite", "bootstrap",
+      "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/String;)Ljava/lang/invoke/CallSite;")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala]: ditto (references to class names should go through a BType)

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala [hunk 3]
```diff
@@ -54,17 +54,20 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
 
   class ProdConsAnalyzer(val methodNode: MethodNode, classInternalName: InternalName) extends AsmAnalyzer(methodNode, classInternalName, new Analyzer(new InitialProducerSourceInterpreter)) with ProdConsAnalyzerImpl
 
+  lazy val lambdaDeserializeBoostrapHandle =
+    new scala.tools.asm.Handle(scala.tools.asm.Opcodes.H_INVOKESTATIC,
+      "scala/runtime/LambdaDeserialize", "bootstrap",
+      "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala]: ditto (references to class names should go through a BType)

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala [hunk 4]
```diff
@@ -77,37 +80,14 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
...
-      mv.visitVarInsn(ASTORE, 1)
-      mv.visitVarInsn(ALOAD, 1)
-      mv.visitFieldInsn(PUTSTATIC, classNode.name, "$deserializeLambdaCache$", mapDesc)
-      mv.visitLabel(l0)
-      mv.visitFieldInsn(GETSTATIC, srLambdaDeserializerRef.internalName, "MODULE$", srLambdaDeserializerRef.descriptor)
-      mv.visitMethodInsn(INVOKESTATIC, jliMethodHandlesRef.internalName, "lookup", nilLookupDesc, false)
-      mv.visitVarInsn(ALOAD, 1)
       mv.visitVarInsn(ALOAD, 0)
-      mv.visitMethodInsn(INVOKEVIRTUAL, srLambdaDeserializerRef.internalName, "deserializeLambda", lookupMapSerlamObjDesc, false)
+      mv.visitInvokeDynamicInsn("lambdaDeserialize", "(Ljava/lang/invoke/SerializedLambda;)Ljava/lang/Object;", lambdaDeserializeBoostrapHandle)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala]: ditto (references to class names should go through a BType)

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala [hunk 5]
```diff
@@ -204,7 +184,8 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
     def visitAnnotationss(annotss: Array[java.util.List[AnnotationNode]]) = if (annotss != null) annotss foreach visitAnnotations
 
     def visitHandle(handle: Handle): Unit = {
-      visitInternalNameOrArrayReference(handle.getOwner)
+      if (handle.getOwner != "scala/runtime/LambdaDeserialize" && handle.getOwner != "scala/runtime/SymbolLiteral")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala]: why is this filter necessary? they are not nested classes

File: src/library/scala/runtime/MethodCache.scala [hunk 6]
```diff
@@ -10,7 +10,8 @@ package scala
 package runtime
 
 
-import java.lang.reflect.{ Method => JMethod }
+import java.lang.invoke.MethodType
+import java.lang.reflect.{Method => JMethod, InvocationTargetException}
```
Comment (lrytz) [inline src/library/scala/runtime/MethodCache.scala]: these changes seem unnecessary

**Comment by DarkDimius** [comment]
> Java interfaces don't allow static fields

Java does allow `public static final` fields in interfaces:

``` java
public interface I {
  static public final Object field = null;
}
```

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala]: Just pushed this. (I had done this already in a separate commit on the branch that I'd cherry-picked this PR from, but forgot to include it.)

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala (hunk 5 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala]: I'll remove this and see what breaks to jog my memory :)

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala [hunk 7]
```diff
@@ -310,6 +310,20 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
...
+          val reflectiveParamsDescriptor = MethodBType(reflectiveParams, ObjectRef).toASMType.getDescriptor
+          mnode.visitInvokeDynamicInsn(qual.symbol.name.encoded, invokedType.getDescriptor, structuralCallSite_BoostrapHandle, reflectiveParamsDescriptor)
+
+        case app @ ApplyDynamic(qual, Literal(Constant(symname: String)) :: Nil) if qual.symbol == Symbol_apply =>
+          val symbol = app.symbol
+          val returnAsmType = typeToBType(symbol.info.resultType).toASMType
+          val invokedType = asm.Type.getMethodType(returnAsmType)
+          val name = "apply"
+          mnode.visitInvokeDynamicInsn(name, invokedType.getDescriptor, symbolLiteralBoostrapHandle, symname)
+
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:321]: This is the ad-hoc part I was referring to; ideally the `ApplyDynamic` tree would have all the info we need to emit the `invokedynamic` instruction.

**Comment by retronym** [comment]
> Java does allow public static final fields in interfaces:

Oh, good to know. I'll keep that in mind as an option (we can simulate a non final field, where needed, with a mutable box).

I think there is an argument that the `invokedynamic` approach still has some merit, as it exposes fewer implementation details as public static fields.

**Comment by retronym** [comment]
I think it is also worthwhile to push forward to support `invokedynamic` generally in the backend, as we can use macros to expose it through to the user level.

**Comment by sjrd** [comment]
Just pointing out that exposing `invokedynamic` at the user level would be a JVM-only feature. I'm pretty sure there's no way to implement such a thing in Scala.js, although I'm not very familiar with the details of `invokedynamic`.

**Comment by retronym** [comment]
> I can't remember why I needed to change visitHandle,

Removing the hack in `visitHandle` led to a test failure in `t6287.scala`.

Turns out that this was related to a problem under multi-run compilation. I'd cached the `asm.Handle` for bootstrap methods in a place in the backend that meant that on a subsequent compilation run, emitting the boostrap method reference didn't populate the `ClassBType` from a compiler `Symbol`. Just before writing the classfile to disk, the backend then attempted to load the class bytes from the compiler classpath instead to determine if it is a nested class. Under the toolbox global, however, we can't (in general) get to classfiles this way, so we hit an assertion failure.

I've pushed an updated version of this PR that moves the lazy vals into the part of the backend that is created for each `Run`.

**Comment by retronym** [comment]
I've pushed another version with the generalization I talked about before. See the new test `test/files/run/indy-via-macro` for an example of a use case that can be solved with this facility.

**Comment by retronym** [comment]
I'll bring this PR back soon.

**Comment by lrytz** [comment]
LGTM!

Very nice work! The generalization is great, and trading `--` in `src/compiler` for `++` in `src/library` is also excellent.

**Comment by retronym** [comment]
Rebased.

**Comment by SethTisue** [comment]
@retronym should we tweak https://github.com/scala/scala/blob/2.12.x/src/library/scala/language.scala#L68-L84 to reflect (no pun intended) this?

**Comment by SethTisue** [comment]
I guess it's fine as is since we still ultimately call the method reflectively.
